src/workerd/io/worker.c++

```diff
@@ -978,6 +979,58 @@ kj::Maybe<kj::String> makeCompatJson(kj::ArrayPtr<kj::StringPtr> enableFlags) {
   return kj::String(json.releaseAsArray());
 }
 
+struct CrossThreadPromiseWaiter {
+  // When a promise is created in a different IoContext, we need to use a
+  // CrossThreadWaitList in order to wait on it. The Waiter instance will
+  // be held on the Promise itself, and will be fulfilled/rejected when the
+  // promise is resolved or rejected. This will signal all of the waiters
+  // from other IoContexts.
+  CrossThreadWaitList waitList;
+  kj::Maybe<jsg::Value> result;
+  jsg::Promise<void> addWaiter(jsg::Lock& js) {
+    return IoContext::current().awaitIo(js, waitList.addWaiter());
+  }
+  void fulfill(jsg::Value value) {
+    result = kj::mv(value);
+    waitList.fulfill();
+  }
+  void reject(kj::Exception ex) {
+    waitList.reject(kj::mv(ex));
+  }
+  CrossThreadPromiseWaiter() = default;
+  KJ_DISALLOW_COPY_AND_MOVE(CrossThreadPromiseWaiter);
+};
+
+CrossThreadPromiseWaiter& getCrossThreadPromiseWaiter(jsg::Lock& js,
+                                                      jsg::V8Ref<v8::Promise>& promise) {
+
+  auto key = v8::Private::ForApi(js.v8Isolate, jsg::v8StrIntern(js.v8Isolate, "waiter"));
+  auto handle = promise.getHandle(js);
+  auto opaqueWaiter = jsg::check(handle->GetPrivate(js.v8Context(), key));
+  if (opaqueWaiter->IsUndefined()) {
+    // Is opaqueWaiter is undefined here, then this is the first time we are waiting on
+    // this promise from a different context. We'll create our CrossThreadWaitList and
+    // set the promise up to resolve it.
+
+    auto waiter = kj::heap<CrossThreadPromiseWaiter>();
+    auto onSuccess = [promise=promise.addRef(js)]
+        (jsg::Lock& js, jsg::Value value) mutable {
+      getCrossThreadPromiseWaiter(js, promise).fulfill(kj::mv(value));
+    };
+
+    auto onFailure = [promise=promise.addRef(js)]
+        (jsg::Lock& js, jsg::Value exception) mutable {
+      getCrossThreadPromiseWaiter(js, promise).reject(js.exceptionToKj(kj::mv(exception)));
```

Comment:
Converting the exception to KJ and back is unfortunate since the recipient is in the same isolate -- letting the JavaScript error object flow through would be best.

Here's an idea: Have the cross-thread promise-fulfiller be of type `<void>`. Have both the success and failure methods simply fulfill the fulfiller. On the receiving end, the `.then()` continuation should _return the original promise_. Since we know the promise is settled at this point, it's now OK to use it directly. This approach correctly propagates either a result value or an error without ever needing to store them anywhere in C++ structures.